File: uninstall.js
var getNode = function(node, fcPath, level){
	if(typeof level === 'undefined') {
		level = 0;
	}
	var words = fcPath.split('/');
	if(words.length <= level){
		return node;
	}
	
	var ret = undefined;
	var children = node.childNodes;
	for (var i = 0; i < children.length; i++){
		if(typeof children[i].getAttribute=== 'undefined'){
			continue;
		}
		if(words[level] == children[i].getAttribute("Name")){
			ret = getNode(children[i], fcPath, level+1);
		}
	}
	
	return ret;
};

var unregisterMacro = function (title, iconPath, macro){
	var shell = new ActiveXObject("WScript.Shell");
	var freeCADHome=shell.ExpandEnvironmentStrings("%APPDATA%") + '\\FreeCAD';
	
	var dom = new ActiveXObject("Msxml2.DOMDocument.6.0");
	var doc = dom.load(freeCADHome + "\\user.cfg");
	if (dom.parseError.errorCode != 0) {
		WScript.Echo("parseError:" + "\n" + dom.parseError.errorCode + "\n" + dom.parseError.reason);
		return;
	}

	var root = dom.documentElement;
	
	// Gets macro path
	var macroPathNode = getNode(root, 'Root/BaseApp/Preferences/Macro/MacroPath');
	var macroPath = (typeof macroPathNode === 'undefined') ? (freeCADHome + '\\') : macroPathNode.text;

	// Unregister macro icon path for ExTools
	var bitmapsNode = getNode(root, 'Root/BaseApp/Preferences/Bitmaps');
	var deleting = [];
	var paths = bitmapsNode.childNodes;
	for (var i =0; i<paths.length; i++){
		if(paths[i].text ==  iconPath){
			deleting.push(i);
		}
	}
	for (var i=deleting.length-1; 0<=i; i--) {
		bitmapsNode.removeChild(bitmapsNode.childNodes[deleting[i]]);
	}

	// Unregister Macro
	var macrosNode = getNode(root, 'Root/BaseApp/Macro/Macros');
	var deleting = [];
	var macros = macrosNode.childNodes;
	for (var i =0; i<macros.length; i++){
		var script = getNode(macros[i], 'Script');
		if(script.text == macro.Script){
			deleting.push(i);
		}
	}
	for (var i=deleting.length-1; 0<=i; i--) {
		macrosNode.removeChild(macrosNode.childNodes[deleting[i]]);
	}
	
	// Unregister to Workbench
	var destWorkbench = getNode(root, 'Root/BaseApp/Workbench/Global');
	if(typeof destWorkbench != 'undefined'){
		var destToolbar = getNode(root, 'Root/BaseApp/Workbench/Global/Toolbar');
		if(typeof destToolbar != 'undefined'){	
			var customs = destToolbar.childNodes;
			var deleting = [];
			for (var i =0; i<customs.length; i++){
				var name = getNode(customs[i], 'Name');
				if(name.text == title){
					deleting.push(i);
				}
			}
			for (var i=deleting.length-1; 0<=i; i--) {
				destToolbar.removeChild(destToolbar.childNodes[deleting[i]]);
			}		
		}
	}

	// Unregister to MainWindow
	var mainWindowToolbar = getNode(root, 'Root/BaseApp/MainWindow/Toolbars');
	if(typeof mainWindowToolbar != 'undefined'){
		var mainWindowToolbarTools = getNode(root, 'Root/BaseApp/MainWindow/Toolbars/' + title);
		if(typeof mainWindowToolbarTools != 'undefined'){
			mainWindowToolbar.removeChild(mainWindowToolbarTools);
		}
	}

	dom.save(freeCADHome + "\\user.cfg");
	return macroPath;
};

(function (){
	var shell = new ActiveXObject("WScript.Shell");
	var freeCADHome=shell.ExpandEnvironmentStrings("%APPDATA%") + '\\FreeCAD';
	
	// register macro
	var title = 'ExTools';
	var iconPath = freeCADHome + '\\ExTools';
	var macroFunctionPolyline = {
		Script : 'FunctionPolyline.FCMacro',
		Menu : 'Function Polyline...',
		Tooltip : 'Draw polyline with function.',
		WhatsThis : 'Draw polyline with function.',
		Statustip : 'Type your function and click "Add"',
		Pixmap : 'FunctionPolyline',
		Accel : 'Alt+F'
	};
	var macroPath = unregisterMacro(title, iconPath, macroFunctionPolyline);
	if(typeof macroPath === 'undefined'){
		return;
	}
	var macroFunctionPolyline = {
		Script : 'ShapeInformation.FCMacro',
		Menu : 'Shape Information...',
		Tooltip : 'Show volume, area and center of mass.',
		WhatsThis : 'Show volume, area and center of mass.',
		Statustip : '',
		Pixmap : 'ShapeInformation',
		Accel : 'Alt+I'
	};
	var macroPath = unregisterMacro(title, iconPath, macroFunctionPolyline);
	if(typeof macroPath === 'undefined'){
		return;
	}

	// Delete macro
	var fso = new ActiveXObject("Scripting.FileSystemObject");
	if(fso.FileExists(freeCADHome + "\\FunctionPolyline.FCMacro")){
		fso.DeleteFile(freeCADHome + "\\FunctionPolyline.FCMacro", true);
	}
	if(fso.FileExists(freeCADHome + "\\ShapeInformation.FCMacro")){
		fso.DeleteFile(freeCADHome + "\\ShapeInformation.FCMacro", true);
	}
	
	// Delete icon
	if(fso.FolderExists(iconPath)){
		fso.DeleteFolder (iconPath, true);
	}

	WScript.Echo('done.');
})();

File: uninstall.py
# -*- coding: utf-8 -*- 

import xml.dom.minidom
import sys
import shutil
import os
import os.path

def getNode(node, fcPath, level=0):
	words = fcPath.split("/");
	if(len(words) <= level):
		return node

	ret = None
	children = node.childNodes
	for node in children:
		if node.nodeType != node.ELEMENT_NODE:
			continue
		if words[level] == node.getAttribute("Name"):
			ret = getNode(node, fcPath, level+1)		
	return ret

def unregisterMacro(title, iconPath, macro):
	if os.name == "nt":
		freeCADHome = os.getenv('APPDATA') +"/FreeCAD"
	elif os.name == "posix":
		freeCADHome = os.path.expanduser("~") +"/.FreeCAD"
	
	dom = xml.dom.minidom.parse(freeCADHome + "/user.cfg")
	root = dom.documentElement
	
	# Gets macro path
	macroPath = ""
	macroPathNode = getNode(root, "Root/BaseApp/Preferences/Macro/MacroPath");
	if macroPathNode != None and macroPathNode.hasChildNodes() and (macroPathNode.childNodes[0].nodeType == macroPathNode.TEXT_NODE):
		macroPath = macroPathNode.childNodes[0].data
	else:	
		macroPath = freeCADHome + "/"

	# Unregister macro icon path for ExTools
	bitmapsNode = getNode(root, "Root/BaseApp/Preferences/Bitmaps")
	existPath = False
	deleting = []
	for path in bitmapsNode.childNodes:
		if path != None and path.hasChildNodes() and (path.childNodes[0].nodeType == path.TEXT_NODE):
			if path.childNodes[0].data == iconPath:
				deleting.append(path)
	for node in deleting:
		bitmapsNode.removeChild(node)
	
	macros = getNode(root, "Root/BaseApp/Macro/Macros")
	deleting = []
	for m in macros.childNodes:
		script = getNode(m, "Script")
		if script != None and script.hasChildNodes() and (script.childNodes[0].nodeType == script.TEXT_NODE):
			if script.childNodes[0].data == macro.Script:
				deleting.append(m)
	for node in deleting:
		macros.removeChild(node)
	
	# Unregister to Workbench
	destWorkbench = getNode(root, "Root/BaseApp/Workbench/Global")
	if destWorkbench != None :
		destToolbar = getNode(root, "Root/BaseApp/Workbench/Global/Toolbar")
		if destToolbar != None:
			deleting = []
			for custom in destToolbar.childNodes:
				name = getNode(custom, "Name")
				if name != None and name.hasChildNodes() and (name.childNodes[0].nodeType == name.TEXT_NODE):
					if name.childNodes[0].data == title:
						deleting.append(custom)
			for node in deleting:
				destToolbar.removeChild(node)
	
	# Unregister to MainWindow
	mainWindowToolbar = getNode(root, "Root/BaseApp/MainWindow/Toolbars")
	if mainWindowToolbar != None:
		mainWindowToolbarTools = getNode(root, "Root/BaseApp/MainWindow/Toolbars/" + title)
		if mainWindowToolbarTools != None:
			mainWindowToolbar.removeChild(mainWindowToolbarTools)
	
	dom.writexml(open(freeCADHome + "/user.cfg","w"))
	return macroPath

def install():
	if os.name == "nt":
		freeCADHome = os.getenv('APPDATA') +"/FreeCAD"
	elif os.name == "posix":
		freeCADHome = os.path.expanduser("~") +"/.FreeCAD"
	
	class Macro:
		def __init__(self):
			self.Script = ""
			self.Menu = ""
			self.Tooltip = ""
			self.WhatsThis = ""
			self.Statustip = ""
			self.Pixmap = ""
			self.Accel = ""

	title = 'ExTools'
	iconPath = freeCADHome + "/ExTools"
	
	macroFunctionPolyline = Macro()
	macroFunctionPolyline.Script = "FunctionPolyline.FCMacro"
	macroFunctionPolyline.Menu = "Function Polyline..."
	macroFunctionPolyline.Tooltip = "Draw polyline with function."
	macroFunctionPolyline.WhatsThis = "Draw polyline with function."
	macroFunctionPolyline.Statustip = 'Type your function and click "Add"'
	macroFunctionPolyline.Pixmap = "FunctionPolyline"
	macroFunctionPolyline.Accel = "Alt+P"
	macroPath = unregisterMacro(title, iconPath, macroFunctionPolyline)
	if macroPath == None:
		sys.exit()
	
	macroShapeInformation = Macro()
	macroShapeInformation.Script = "ShapeInformation.FCMacro"
	macroShapeInformation.Menu = "Shape Information..."
	macroShapeInformation.Tooltip = "Show volume, area and center of mass."
	macroShapeInformation.WhatsThis = "Show volume, area and center of mass."
	macroShapeInformation.Statustip = ""
	macroShapeInformation.Pixmap = "ShapeInformation"
	macroShapeInformation.Accel = "Alt+I"
	macroPath = unregisterMacro(title, iconPath, macroShapeInformation)
	if macroPath == None:
		sys.exit()
	
	#  Delete macro
	if os.path.exists(macroPath + "/FunctionPolyline.FCMacro"):
		os.remove(macroPath + "/FunctionPolyline.FCMacro")
	if os.path.exists(macroPath + "/ShapeInformation.FCMacro"):
		os.remove(macroPath + "/ShapeInformation.FCMacro")

	# Delete icon
	if os.path.exists(iconPath):
		shutil.rmtree(iconPath)

if __name__ == "__main__":
	install()
	print("done.")